### PR TITLE
drivers: can: add start and stop CAN controller API functions

### LINKED ISSUE #45304: drivers: can: CAN interfaces are brought up with default bitrate at boot, causing error frames if bus bitrate differs
**Describe the bug**
Currently, the CAN drivers start with a default bitrate (or timing) specified using devicetree and start the CAN controller in their respective `init()` functions.

This is fine for CAN nodes using only one fixed bitrate, but if the bitrate is set by the user (e.g. via a DIP-switch or other HMI which is very common), the CAN driver will still initialise with the default timing/bitrate at boot and use this until the application has determined the requested bitrate/timing and set it using `can_set_bitrate()`/`can_set_timing()`.

During this period, the CAN node will potentially destroy valid CAN frames on the CAN bus (which is using the soon-to-be-set-by-the-application bitrate) by sending error frames.

**Expected behavior**
The CAN controller drivers should not automatically start the CAN controller at boot time, but instead provide an API for the application to start/stop the CAN controller.

**Impact**
Interruptions to the ongoing CAN bus traffic when a Zephyr-based CAN node connected to the bus is (re-)booted.

**Additional context**
The Linux kernel CAN drivers require all configuration (bitrate/timing, mode, etc.) to take place before bringing the CAN interface "up". Reconfiguration requires bringing the CAN interface "down" first.

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Up until now, the Zephyr CAN controller drivers set a default bitrate (or timing) specified via devicetree and start the CAN controller in their respective driver initialization functions.
    
This is fine for CAN nodes using only one fixed bitrate, but if the bitrate is set by the user (e.g. via a DIP-switch or other HMI which is very common), the CAN driver will still initialise with the default bitrate/timing at boot and use this until the application has determined the requested bitrate/timing and set it using can_set_bitrate()/can_set_timing().
    
During this period, the CAN node will potentially destroy valid CAN frames on the CAN bus (which is using the soon-to-be-set-by-the-application bitrate) by sending error frames. This causes interruptions to the ongoing CAN bus traffic when a Zephyr-based CAN node connected to the bus is (re-)booted.
    
Instead, require all configuration (setting bitrate, timing, or mode) to take place when the CAN controller is stopped. This maps nicely to entering "reset mode" (called "configuration mode" or "freeze mode" for some CAN controller implementations) when stopping and exiting this mode when starting the CAN controller.
    
Fixes: #45304

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
These driver and API changes were verified on the following boards:
- NXP FlexCAN driver:
  - `twr_ke18f`
  - `mimxrt1024_evk`
  - `frdm_k64f`
- STM32 bxCAN:
  - `stm32f3_disco`
- Bosch M_CAN:
  - `lpcxpresso55s16`
  - `nucleo_g474re`
- SJA1000:
  - `olimex_esp32_evb`
- Microchip MCP2515:
  - `frdm_k64f` + `keyestudio_can_bus_ks0411` shield
- CAN loopback:
  - `native_posix`
  - `native_posix_64`
- CAN Native Posix Linux with virtual CAN device:
  - `native_posix`
  - `native_posix_64`

The changes to the Renesas R-Car CAN driver was not verified on hardware as I do not have access to any such development boards.

**Comment by aaillet** [comment]
> The changes to the Renesas R-Car CAN driver were not verified on hardware as I do not have access to any such development boards.

@pmarzin will review the R-Car driver and do some tests on dev boards, please wait for it before merging these changes !

**Comment by mmahadevan108** [comment]
@henrikbrixandersen I have cleared the milestone field as we are past the feature freeze date for the 3.2 release. Kindly pick a different one for a future release.

**Comment by mmahadevan108** [comment]
@henrikbrixandersen Readded milestone as this is a bug fix. Sorry I missed that.

**Review by martinjaeger — COMMENTED** [review]
Generally looks good to me except for the two minor comments.

However, I was wondering if the PR actually solves the original issue with the controller being started automatically with the default bitrate for all drivers.

As far as I can see, almost all `can_init` functions are left untouched. In the case of e.g. STM32 the init function already sets the controller into normal operation mode, so I guess it would produce error frames if it receives messages with wrong bitrate (didn't test it in hardware, though). Should some parts of the `can_init` functions not be moved to `can_start`?

I see that the transceivers are not yet enabled, but not all boards will have a transceiver with `STB` pin, which is disabled during start-up.

File: drivers/can/can_loopback.c [hunk 1]
```diff
@@ -235,30 +270,46 @@ static int can_loopback_set_mode(const struct device *dev, can_mode_t mode)
...
 	return 0;
 }
 #endif /* CONFIG_CAN_FD_MODE */
 
 static int can_loopback_get_state(const struct device *dev, enum can_state *state,
 				  struct can_bus_err_cnt *err_cnt)
 {
+	struct can_loopback_data *data = dev->data;
+
 	ARG_UNUSED(dev);
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply drivers/can/can_loopback.c]: Fixed, thanks.

File: drivers/can/can_sja1000.c [hunk 2]
```diff
@@ -683,17 +742,13 @@ int can_sja1000_init(const struct device *dev)
...
 	can_sja1000_write_reg(dev, CAN_SJA1000_OCR, config->ocr);
 
-	/* Clear error counters */
-	can_sja1000_write_reg(dev, CAN_SJA1000_RXERR, 0);
-	can_sja1000_write_reg(dev, CAN_SJA1000_TXERR, 0);
-
-	/* Clear error capture */
-	(void)can_sja1000_read_reg(dev, CAN_SJA1000_ECC);
+	/* Clear error counters and error capture */
+	can_sja1000_clear_errors(dev);
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply drivers/can/can_sja1000.c:746]: I kept the comment since the flow of comments match the initialization sequence described in the data sheet.

**Comment by henrikbrixandersen** [comment]
> However, I was wondering if the PR actually solves the original issue with the controller being started automatically with the default bitrate for all drivers.

It does. I have verified this on real hardware.

> As far as I can see, almost all `can_init` functions are left untouched. In the case of e.g. STM32 the init function already sets the controller into normal operation mode, so I guess it would produce error frames if it receives messages with wrong bitrate (didn't test it in hardware, though). Should some parts of the `can_init` functions not be moved to `can_start`?

Generally, the `init()` functions call the `set_mode()` functions, which were putting the CAN controllers into operational mode. The `set_mode()` functions have been verified to not do this any longer.

> I see that the transceivers are not yet enabled, but not all boards will have a transceiver with `STB` pin, which is disabled during start-up.

I have tested all but the Renesas R-Car driver without any actively controlled CAN transceivers and verified the correct behaviour.

**Comment by martinjaeger** [comment]
> Generally, the `init()` functions call the `set_mode()` functions, which were putting the CAN controllers into operational mode. The `set_mode()` functions have been verified to not do this any longer.

Ok, I missed that the `leave_init_mode()` is not called anymore during init, as it's not part of `set_mode()` anymore. So it's all good.

**Review by martinjaeger — APPROVED** [review]

**Comment by pmarzin** [comment]
@henrikbrixandersen Tests are good on RCAR can driver (tested in loopback with twister and with physical link).
It's good for with @aaillet :)

**Review by aaillet — APPROVED** [review]

**Pull request merged**